AppCenterAuth/AppCenterAuth/MSAuth.m

```diff
@@ -269,44 +278,66 @@ - (MSAuthConfigIngestion *)ingestion {
-                                                               withData:data
-                                                             atomically:YES
-                                                         forceOverwrite:YES];
-
-                // Store eTag only when the configuration file is created successfully.
-                if (configUrl) {
-                  NSString *newETag = [MSHttpIngestion eTagFromResponse:response];
-                  if (newETag) {
-                    [MS_USER_DEFAULTS setObject:newETag forKey:kMSAuthETagKey];
-                  }
-                } else {
-                  MSLogWarning([MSAuth logTag], @"Couldn't create Auth config file.");
-                }
-                @synchronized(self) {
-                  self.authConfig = config;
-
-                  // Reinitialize client application.
-                  [self configAuthenticationClient];
+  [self.ingestion
+              sendAsync:nil
+                   eTag:eTag
+      completionHandler:^(__unused NSString *callId, NSHTTPURLResponse *response, NSData *data, __unused NSError *error) {
+        @synchronized (self) {
+          BOOL continueSignIn = self.signInCompletionHandler && !self.clientApplication;
+          self.signInShouldWaitForConfig = NO;
+          MSAuthConfig *config = nil;
+          if (response.statusCode == MSHTTPCodesNo304NotModified) {
+            MSLogInfo([MSAuth logTag], @"Auth config hasn't changed.");
+
+            // At this point, the expectation is that we should not have any pending sign-ins, because
+            // the configuration already exists. If we have a pending sign-in, this will trigger an error.
+            [self callCompletionHandler:self.signInCompletionHandler
+                          withErrorCode:MSACAuthErrorSignInConfigNotValid
+                             andMessage:@"There was no auth config but the server returned 304 (not modified)."];
+          } else if (response.statusCode == MSHTTPCodesNo200OK) {
+            config = [self deserializeData:data];
+            if ([config isValid]) {
+              NSURL *configUrl = [MSUtility createFileAtPathComponent:[self authConfigFilePath]
+                                                             withData:data
+                                                           atomically:YES
+                                                       forceOverwrite:YES];
+
+              // Store eTag only when the configuration file is created successfully.
+              if (configUrl) {
+                NSString *newETag = [MSHttpIngestion eTagFromResponse:response];
+                if (newETag) {
+                  [MS_USER_DEFAULTS setObject:newETag forKey:kMSAuthETagKey];
                 }
               } else {
-                MSLogError([MSAuth logTag], @"Downloaded auth config is not valid.");
+                MSLogWarning([MSAuth logTag], @"Couldn't create Auth config file.");
+              }
+              @synchronized(self) {
```

Comment:
There's a `@synchronized` block in line 288, why do we need a nested `@synchronized` here?